runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/status/BeamWorkerStatusGrpcService.java

```diff
@@ -92,90 +92,97 @@ public void close() throws Exception {
+              } else {
+                existingClientFuture.complete(fnApiStatusClient);
+                return existingClientFuture;
+              }
+            } catch (IOException | InterruptedException | ExecutionException e) {
+              LOG.warn("Error closing worker status client", e);
+            }
+          }
+          return CompletableFuture.completedFuture(fnApiStatusClient);
+        });
     return fnApiStatusClient.getResponseObserver();
   }
 
   /**
-   * Get the latest SDK worker status from the client's corresponding SDK Harness.
+   * Get the latest SDK worker status from the client's corresponding SDK harness.
    *
    * @param workerId worker id of the SDK harness.
    * @return {@link CompletableFuture} of WorkerStatusResponse from SDK harness.
    */
-  public CompletableFuture<WorkerStatusResponse> getWorkerStatus(String workerId) {
+  public String getSingleWorkerStatus(String workerId, long timeout, TimeUnit timeUnit) {
+    if (!connectedClient.containsKey(workerId)) {
+      return "Error: Not connected.";
+    }
     try {
-      return getStatusClient(workerId).getWorkerStatus();
+      return getWorkerStatus(workerId).get(timeout, timeUnit);
     } catch (InterruptedException | ExecutionException | TimeoutException e) {
-      if (e instanceof InterruptedException) {
-        Thread.currentThread().interrupt();
-      }
-      CompletableFuture<WorkerStatusResponse> error = new CompletableFuture<>();
-      error.completeExceptionally(e);
-      return error;
+      return handleAndReturnExceptionResponse(e);
     }
   }
 
   /**
-   * Get all the statuses from all connected SDK harnesses within specified timeout.
+   * Get all the statuses from all connected SDK harnesses within specified timeout. Any errors
+   * getting status from the SDK harnesses will be returned in the map.
    *
-   * @param timeout max time waiting for the response from all SDK harness.
+   * @param timeout max time waiting for the response from each SDK harness.
    * @param timeUnit timeout time unit.
-   * @return All the statuses in a Map which key is the SDK harness id and value is the status
-   *     content.
+   * @return All the statuses in a map keyed by the SDK harness id.
    */
   public Map<String, String> getAllWorkerStatuses(long timeout, TimeUnit timeUnit) {
     // return result in worker id sorted map.
     Map<String, String> allStatuses = new ConcurrentSkipListMap<>(Comparator.naturalOrder());
 
-    List<CompletableFuture<WorkerStatusResponse>> responses = new ArrayList<>();
-    for (String id : connectedClient.keySet()) {
-      responses.add(
-          getWorkerStatus(id)
-              .whenComplete(
-                  (response, ex) -> {
-                    allStatuses.put(
-                        id, ex != null ? DEFAULT_ERROR_RESPONSE : getStatusErrorOrInfo(response));
-                  }));
-    }
+    connectedClient
+        .keySet()
+        .parallelStream()
```

Comment:
This can throw a [ConcurrentModificationException](https://docs.oracle.com/javase/7/docs/api/java/util/ConcurrentModificationException.html) if the parallel stream is over a connectedClient that is being mutated while your iterating over it.